### PR TITLE
test: Add multidimensional interop test

### PR DESCRIPTION
**Pull request by MarcoPolo**
Implements: https://github.com/libp2p/test-plans/blob/master/multidim-interop/README.md

### CODE REVIEW

**Comment by MarcoPolo** [comment]
Future todos:
- [ ] Figure out why firefox webrtc isn't working
- [ ] Figure out why webkit tests fail to create a libp2p node

**Comment by MarcoPolo** [comment]
@ckousik I'm seeing a failure with firefox -> webrtc -> rust (Chrome works fine).  https://github.com/libp2p/js-libp2p/actions/runs/4369651820/jobs/7643671824#step:5:5535 " localCert.getFingerprints is not a function". Have you seen that before? https://github.com/libp2p/js-libp2p-webrtc/blob/4c8806c6d2a1a8eff48f0e2248203d48bd84c065/src/transport.ts#L215

**Comment by MarcoPolo** [comment]
I don't know why windows is failing. It seems like it's failing in other PRs as well, so probably not related to these changes

**Comment by ckousik** [comment]
> @ckousik I'm seeing a failure with firefox -> webrtc -> rust (Chrome works fine). https://github.com/libp2p/js-libp2p/actions/runs/4369651820/jobs/7643671824#step:5:5535 " localCert.getFingerprints is not a function". Have you seen that before? https://github.com/libp2p/js-libp2p-webrtc/blob/4c8806c6d2a1a8eff48f0e2248203d48bd84c065/src/transport.ts#L215

That's odd, since `getFingerprints` is required in w3c https://w3c.github.io/webrtc-pc/#rtccertificate-interface . Looking into it now.

**Comment by MarcoPolo** [comment]
This ready for review. This could be faster, our build of head takes 10m. I haven't spent any time investigating this, but we could probably speed this up.

**Comment by p-shahi** [comment]
@achingbrain Can we review and get this in this week (meant to follow up on this in triage today)?

**Review by achingbrain — APPROVED** [review]
LGTM.

We should add the `interop` folder to the list of files that get linted to keep everything consistent.  At the moment there are some rules violations but we can fix them when they actually start to break something.

Adding 10m to the build time seems a bit excessive - can it be faster?

**Comment by MarcoPolo** [comment]
Per sync discussion:

@achingbrain is hesistant to merge this while it takes >10min. If I bring it down to <10min, we'll merge.

**Comment by MarcoPolo** [comment]
Alright, I spent a good bit of time optimizing this and the tests run in 15 min. I think this is as good as we can get it without reworking the test runner itself. (unfortunately docker run startup time is the limiting factor, with each test taking about 2s to run).

This could improve a bit if we make the rust images smaller, but I wouldn't expect too much.

@achingbrain can I get another approval?

**Comment by thomaseizinger** [comment]
> This could improve a bit if we make the rust images smaller, but I wouldn't expect too much.

Tracking this here: https://github.com/libp2p/rust-libp2p/issues/3881

**Comment by MarcoPolo** [comment]
All tests pass! @achingbrain any objections to merging?

**Comment by thomaseizinger** [comment]
@MarcoPolo Do I understand correctly that with this PR js-libp2p will use the same structure of just pointing to a commit-hash in the test-plans repo?

Much in favor of adopting that!

File: .github/workflows/interop-test.yml
Comment (maschad) [inline .github/workflows/interop-test.yml:22]: Why is this commented out? I think that if we want tests to be reproducible for a given commit this would be required otherwise we are vulnerable to dep breakages

File: interop/Dockerfile
Comment (maschad) [inline interop/Dockerfile:22]: I am guessing this is irrelevant now, if so we should remove.

**Comment by MarcoPolo** [comment]
> Do I understand correctly that with this PR js-libp2p will use the same structure of just pointing to a commit-hash in the test-plans repo?


Yes

**Comment by MarcoPolo** [comment]
@achingbrain this is blocking more interop tests in test plans and webrtc interop tests. Since this is only test code, I'll assume your previous approval is still valid unless you say otherwise within the next ~24h.

File: interop/test/ping.spec.ts
Comment (achingbrain) [inline interop/test/ping.spec.ts:105]: Fixed in https://github.com/libp2p/js-libp2p/pull/1752

File: interop/test/ping.spec.ts
Comment (achingbrain) [inline interop/test/ping.spec.ts:161]: If this isn't part of the test, maybe do this in an `afterEach` block?

File: interop/test/ping.spec.ts
Comment (achingbrain) [inline interop/test/ping.spec.ts:34]: Can all this config be built in a separate file and invoked in a `beforeEach` block?

It's very verbose and doesn't seem like it's part of the test.

File: interop/test/ping.spec.ts
Comment (achingbrain) [inline interop/test/ping.spec.ts:147]: This will always take the maximum amount of time, is there no way to wait for whatever it is we want to happen?  Otherwise these tests may spend a lot of time doing nothing.

File: interop/test/ping.spec.ts
Comment (achingbrain) [inline interop/test/ping.spec.ts:126]: Conditional logic in tests is always raises an eyebrow as you can end up not testing what you think you are testing.

Can this be split into two separate tests?

For example the dialer test could skip itself if the node isn't the dialer on this run.

File: interop/.aegir.js
Comment (achingbrain) [inline interop/.aegir.js:80]: Do we care about the bundle size here?  Probably not?  This block can be removed.

File: interop/package.json
Comment (achingbrain) [inline interop/package.json:37]: Please align aegir version with the main repo.

File: interop/package.json
Comment (achingbrain) [inline interop/package.json:39]: Aegir comes with standard and typescript, we probably don't need them here?

**Review by achingbrain — CHANGES_REQUESTED** [review]

File: interop/test/ping.spec.ts
Comment (MarcoPolo) [reply interop/test/ping.spec.ts:147]: This is on purpose. The test runner will shutdown the listener when the dial completes. See https://github.com/libp2p/test-plans/tree/master/multidim-interop#listener

File: interop/test/ping.spec.ts
Comment (MarcoPolo) [reply interop/test/ping.spec.ts:126]: Why? Every other [implementation](https://github.com/libp2p/go-libp2p/blob/master/test-plans/cmd/ping/main.go#L161) [does](https://github.com/libp2p/rust-libp2p/blob/master/interop-tests/src/bin/ping.rs#L127) [it](https://github.com/libp2p/test-plans/blob/master/multidim-interop/impl/nim/mainv1.nim#L70) [this](https://github.com/MarcoPolo/zig-libp2p/blob/main/interop/main.zig#L118) way.

File: interop/.aegir.js
Comment (MarcoPolo) [reply interop/.aegir.js:80]: Vestigal. Will remove.

File: interop/test/ping.spec.ts
Comment (MarcoPolo) [reply interop/test/ping.spec.ts:34]: I don't see the benefit of having a separate file, but I can move to a beforeEach.

File: interop/package.json
Comment (MarcoPolo) [reply interop/package.json:43]: If I recall correctly, we needed a newer browser for some test. This shouldn't be needed after https://github.com/hugomrdias/playwright-test/pull/528/files.

If we package.json supported comments it would have been easier to capture the rational here.

File: interop/test/ping.spec.ts
Comment (achingbrain) [reply interop/test/ping.spec.ts:39]: Presumably these tests will test more than just ping in the future, at that point we're going to want to set up a libp2p node in more than one place, hence the external file.

We can make that refactor when we need it, but also this is setup code and not test code so it shouldn't be in the `it` function.

File: interop/test/ping.spec.ts
Comment (achingbrain) [reply interop/test/ping.spec.ts:139]: Looking at the other implementations, none of them are written using a test framework so I don't think they are an example to follow here.

Comment (MarcoPolo) [reply interop/test/ping.spec.ts:139]: Maybe we shouldn’t use the test framework? I only used it because it was convenient. Happy to not use it.

File: interop/test/ping.spec.ts
Comment (MarcoPolo) [reply interop/test/ping.spec.ts:39]: This is testing whether two libp2p nodes can communicate. A future test may look similar, but I don’t think it makes sense to refactor until we have that future test.

File: interop/test/ping.spec.ts
Comment (MarcoPolo) [reply interop/test/ping.spec.ts:139]: Do you feel strongly enough against the test framework as to request changes before merging?

I don’t think we should deviate from what the four other implementations do just follow the convention on this test framework, so I would suggest keeping the conditional as is, and , if requested, moving off the test framework. 

I would be against splitting this into two tests.

Comment (achingbrain) [reply interop/test/ping.spec.ts:139]: I think using a test framework to write tests is the right thing to do, as is integrating with the existing tooling of a project because it makes the code instantly familiar to other developers so I think this is the right decision.

I think regardless of framework tests have a certain structure (setup, exercise, verify, teardown), duplication should be minimised and that conditional logic in tests means that later down the line when something goes wrong the author will have to think about which branch is being executed which is a code smell and should be avoided.

I also think that the first test is the most important since people will use it as a starting point for subsequent tests and am somewhat confused as to why you are pushing back on this so hard.

**Comment by maschad** [comment]
> I've split up the test @achingbrain, but what's annoying now is that one or the other test will always be skipped but it will still spin up and shutdown a libp2p node because of beforeEach and afterEach. Any suggestions? We might be able to replace it(... with it.skip(... dynamically and thus skip the test at the higher level, but that seems a bit hacky.

I agree @MarcoPolo it seems to get a bit messy going that route. What do you think about splitting the tests into two separate directories i.e. one for dialler tests and one for listener tests (albeit they are both just one test each at the moment) and we can use just use an `npm` script like 

```
"test": if [ \"$is_dialer\" = \"true\" ]; then aegir test -t node -f \".dist/test/dialer/*.{node,spec}.js\"; else aegir test -t node -f \".dist/test/listener/*.{node,spec}.js\"; fi
```

I know we have been going back and forth on this for a while so if you and @achingbrain agree then I could make these changes so that we can land this.

**Comment by MarcoPolo** [comment]
> > I've split up the test @achingbrain, but what's annoying now is that one or the other test will always be skipped but it will still spin up and shutdown a libp2p node because of beforeEach and afterEach. Any suggestions? We might be able to replace it(... with it.skip(... dynamically and thus skip the test at the higher level, but that seems a bit hacky.
> 
> I agree @MarcoPolo it seems to get a bit messy going that route. What do you think about splitting the tests into two separate directories i.e. one for dialler tests and one for listener tests (albeit they are both just one test each at the moment) and we can use just use an `npm` script like
> 
> ```
> "test": if [ \"$is_dialer\" = \"true\" ]; then aegir test -t node -f \".dist/test/dialer/*.{node,spec}.js\"; else aegir test -t node -f \".dist/test/listener/*.{node,spec}.js\"; fi
> ```

I'm strongly against separate files for this test. I also don't like using bash scripting as the way to check which test to run. Why are we introducing bash into this when we're already working with TypeScript?

**Comment by maschad** [comment]
> I'm strongly against separate files for this test. 

why is that? As it is now [they are two separate tests](https://github.com/libp2p/js-libp2p/pull/1615/files#diff-024eabde3750f3bf282cd9a642bd16ba0357c8c0b3110ffa4c78bfdd60f9ed2fR135) in one file, I think that having them in separate files improves the modularity and reporting (as you wouldn't see the logs of the test being skipped)

>  I also don't like using bash scripting as the way to check which test to run. Why are we introducing bash into this when we're already working with TypeScript?

I can appreciate that but I think they serve two different purposes in this case. I would say it's more common to see bash scripting accessing environment variables than Typescript. I would also say it's more intuitive than [the dynamic way of skipping tests](https://github.com/libp2p/js-libp2p/pull/1615/files#diff-024eabde3750f3bf282cd9a642bd16ba0357c8c0b3110ffa4c78bfdd60f9ed2fR135)

Regardless, I wouldn't let either of those concerns block this from being landed and I see that [tests pass now](https://github.com/libp2p/js-libp2p/actions/runs/5063541633/jobs/9090264246?pr=1615)  🚀. We can always optimize in a follow up PR 😄

**Comment by MarcoPolo** [comment]
> why is that? As it is now [they are two separate tests](https://github.com/libp2p/js-libp2p/pull/1615/files#diff-024eabde3750f3bf282cd9a642bd16ba0357c8c0b3110ffa4c78bfdd60f9ed2fR135) in one file, I think that having them in separate files improves the modularity and reporting (as you wouldn't see the logs of the test being skipped)

Having more files here doesn't improve modularity, it adds indirection. You may as well run the test command and grep for only the tests you want to run.

Is it common to see Bash scripting used in `npm test`? I've done it a couple of times, but only as a last resort.

**Review by maschad — APPROVED** [review]

**Comment by MarcoPolo** [comment]
> You may as well run the test command and grep for only the tests you want to run.

Sorry this wasn't clear. I was using this an example of what you wouldn't want to do.

**Comment by p-shahi** [comment]
Poking in here: were the discussions mostly resolved i.e. is this guy almost there to getting merged?

**Comment by achingbrain** [comment]
I meant something like:

```js
it('should be a listener', function () {
  if (is_dialer) {
    return this.skip()
  }

  // listener test code here
})
```

The dynamic skip:

```js
(isDialer ? it.skip : it)('should listen for ping', async () => {
```

is reminiscent of [what we did](https://github.com/ipfs/js-ipfs/blob/master/packages/interface-ipfs-core/src/utils/mocha.js#L98) for the js-ipfs interface tests.  It worked, though typing it all was a little hairy.

I think one file for the ping test is ok, definitely don't want to be debugging bash scripts to figure out why tests aren't running.

I think we should get this in and we can finesse it in follow PRs.

**Review by achingbrain — APPROVED** [review]

**Pull request merged**
